### PR TITLE
WIP: Example report exercise steps

### PR DESCRIPTION
**Pull request by LuckyJosh**
Still Work in Progress but only one or two step are missing (for a checklist of the different steps see #419, which I do have to update).

I tried a few methods to reduce the amount of duplication, e.g. using diff and patch or branches in a git repo to `switch` to the next step. In both versions you had to fix the same error in multiple files (the same file in multiple branches) which also felt error prone and inconvenient.

So I came up with this solution:
Each file has a template file on which each version of this file is based upon. 
For example all `auswertung.py` files in each of the 10 step directories are generated from 
one template file `auswertung_py_template`[1]

The template 'files-format' adds just one comment at the end of each line, that contains the information 
in which step the respective line is included. The 'range'-syntax is inspired by the latex-beamer overlay syntax 
(a python slice like `2:4` would have been confusing, since these ranges are inclusive on both ends).

An example from the `auswertung_py_template`:
```python
# padding so that the copy button does not overlap the code
#
#
# Erklärung:                                                                                       # <2-2>
# Die Daten aus Dateien mit mehreren spalten muss man in einer extra Zeile skalieren               # <2-2>
# Masse und Umfang der Kugel                                                                       # <2-10>
m_b, u_b = np.genfromtxt("data/Messwerte_Kugel.txt", delimiter=",", unpack=True)                   # <2-4>
m_b, m_b_unc, u_b, u_b_unc = np.genfromtxt("data/Messwerte_Kugel.txt", delimiter=",", unpack=True) # <5-10>
m_b = m_b/1000 # kg                                                                                # <2-4>
u_b = u_b/100 # m                                                                                  # <2-4>
m_b = unc.ufloat(m_b,m_b_unc)/1000 # kg                                                            # <5-10>
u_b = unc.ufloat(u_b,m_b_unc)/100 # m                                                              # <5-10>
```
The first two lines in this snippet are only apearing in `step-2/auswertung.py`,  the third one in all ten steps.
In step-5 uncertainties are included and so from this step onwards the loading of the datafiles has to be changed.

Before using this custom 'file format' I tried `json` for the template files, but that was ways less readable than
this minimal version.

In addition to adding the 10th step I probably could  clean up the `Makefile` to some extend.

[1] the templates don't have a file ending to prevent formatters from destroying the alignment of the comments

### CODE REVIEW

**Comment by chrbeckm** [comment]
For this year, we could tell them to do it in their version of the template repo.

File: exercises-toolbox/8-all/example-report/Makefile [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+steps_without_unc = 1 2 3 4 
+steps_with_unc = 5 6 7 8 9 10
+steps_with_latex = 7 8 9 10
+
+
+steps = $(steps_without_unc) $(steps_with_unc)
+
+template_taskfiles = $(foreach step, $(steps), $(addprefix templates/, aufgabe-step-$(step).txt))
+template_bibfiles = $(addprefix templates/, lit.bib programme.bib)
+template_setup_image = templates/experimental-setup.png
```
Comment (chrbeckm) [inline exercises-toolbox/8-all/example-report/Makefile:21]: `template_setup_image = templates/**versuchsaufbau**.png`

File: exercises-toolbox/8-all/example-report/Makefile [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+$(target_texfiles_durchfuehrung) \
+$(target_texfiles_theorie) \
+$(target_texfiles_auswertung) \
+$(target_texfiles_diskussion) \
+$(target_bibfiles)\
+$(target_setup_image)
+
+
+build/report-example-step-%/aufgabe.txt &: $(template_taskfiles) | $(solution_dirs)
+	cp $(word $*, $(template_taskfiles)) build/report-example-step-$*/aufgabe.txt
```
Comment (chrbeckm) [inline exercises-toolbox/8-all/example-report/Makefile:69]: Leads to `aufgabe.txt` dirs on my system.

**Comment by LuckyJosh** [comment]
> For this year, we could tell them to do it in their version of the template repo.

Fine by me. :+1:

Using the full template repo would make (at least) step 10 more or less obsolete

I thought the template might be less overwhelming if the participants building it up by themself.

**Comment by LuckyJosh** [comment]
Everything except for step-11 should be fine and running with `make`

**Comment by chrbeckm** [comment]
> > For this year, we could tell them to do it in their version of the template repo.
> 
> Fine by me. 👍
> 
> Using the full template repo would make (at least) step 10 more or less obsolete
> 
> I thought the template might be less overwhelming if the participants building it up by themself.

I thought about
1. Using the template repo to create their own repo
2. create `v16516` dir in that directory, maybe by copying vXXX
3. start with step-02 from this exercise

**Comment by LuckyJosh** [comment]
Done with the build process of all 11 steps, missing:
-  the task descriptions have to be written.
-  hooking into the makefile change of the repo

**Comment by aknierim** [comment]
Is it expected behaviour that the full report is built in both step 1 and step 11?

**Comment by LuckyJosh** [comment]
> Is it expected behaviour that the full report is built in both steps 1 and 11?

Yes, to have a reference from the start.

**Comment by aknierim** [comment]
Okay, makes sense

**Comment by chrbeckm** [comment]
Starting with step 07, I only see
```
# Beispiel Protokoll

Einleitung:
Um dieser Aufgabe sinnvoll folgen zu können sollte diese von Anfang an bearbeitet werden. Folgt den 
Anweisungen in den 'aufgabe.txt' Dateien in den Unterordnern des Ordners 'example-report'.
```
in `aufgabe.txt`.

**Comment by aknierim** [comment]
This is expected behaviour as the instructions seem to be placholders from 7 onward. See `8-all/example-report/templates/aufgabe-step-*.txt`

**Comment by aknierim** [comment]
That is something that has to be fixed of course.

**Comment by LuckyJosh** [comment]
Yes I added the missing task descriptions 7-11 :+

**Review by chrbeckm — APPROVED** [review]

**Pull request merged**